File: java/order-service/src/main/java/net/multi/language/order/service/OrderService.java
package net.multi.language.order.service;

import net.multi.language.order.outbound.port.OutboundMessagingPort;
import net.multi.language.common.dto.order.OrderDTO;
import net.multi.language.common.message.command.order.ChargeOrderCommand;
import net.multi.language.common.message.command.order.CreateOrderCommand;
import net.multi.language.common.message.event.order.OrderCreatedEvent;
import net.multi.language.order.domain.model.Order;
import net.multi.language.order.inbound.port.InboundMessagingPort;
import net.multi.language.order.inbound.port.InboundRestPort;
import net.multi.language.order.outbound.port.OutboundPersistencePort;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.client.RestClient;

import java.util.Random;
import java.util.concurrent.CompletableFuture;

@Service
public class OrderService implements InboundRestPort, InboundMessagingPort {

    private static final Logger LOGGER = LoggerFactory.getLogger(OrderService.class);
    private static final Random RANDOM = new Random(100);

    private final OutboundMessagingPort outboundMessagingPort;
    private final OutboundPersistencePort outboundPersistencePort;

    @Value("${billing.service.endpoint}")
    private String billingServiceEndpoint;

    @Autowired
    public OrderService(final OutboundMessagingPort outboundMessagingPort,
                        final OutboundPersistencePort outboundPersistencePort) {
        this.outboundMessagingPort = outboundMessagingPort;
        this.outboundPersistencePort = outboundPersistencePort;
    }

    // creating an order received from a REST endpoint (UI, testing app etc)
    @Override
    @Transactional
    public void createOrder(final OrderDTO orderDTO) {
        final long customerId = orderDTO.getCustomerId();
        final long orderId = saveOrder(convertDTOIntoOrder(orderDTO));

        LOGGER.info("Creating an order for the customer with the ID {}, for {} items...", customerId,
                orderDTO.getOrderItems().size());
        final double orderTotal = orderDTO.getOrderTotal();

        CompletableFuture.runAsync(() -> publishChargeOrder(customerId, orderId, orderTotal))
                         .thenRunAsync(() -> publishOrderCreatedEvent(customerId, orderId));
    }

    @Override
    public void chargeOrder(long customerId, long orderId) {
        final RestClient restClient = RestClient.create();
        final ResponseEntity<String> response = chargeOrder(customerId, orderId, restClient);
        LOGGER.info("Got the response: {}", response.getBody());
    }

    private @NotNull ResponseEntity<String> chargeOrder(long customerId, long orderId, RestClient restClient) {
        return restClient.post()
                         .uri(billingServiceEndpoint + "/" + customerId + "/" + orderId)
                         .retrieve()
                         .toEntity(String.class);
    }

    // creating an order received from a messaging endpoint (upstream system, 3rd party application etc)
    @Override
    @Transactional
    public void createOrder(final CreateOrderCommand createOrderCommand) {
        final long customerId = createOrderCommand.getCustomerId();
        LOGGER.info("Creating an order for the product '{}', for the customer with the ID {}...",
                createOrderCommand.getProductName(), customerId);

        final long orderId = saveOrder(convertCommandIntoOrder(createOrderCommand));
        final double orderTotal = createOrderCommand.getOrderTotal();

        publishChargeOrder(customerId, orderId, orderTotal);
        publishOrderCreatedEvent(customerId, orderId);
    }

    private void publishChargeOrder(final long customerId, final long orderId, final double orderTotal) {
        outboundMessagingPort.publishChargeOrderCommand(
                new ChargeOrderCommand(getNextMessageId(), customerId, orderId, orderTotal,
                        ChargeOrderCommand.Currency.EUR));
    }

    private void publishOrderCreatedEvent(final long customerId, final long orderId) {
        outboundMessagingPort.publishOrderCreatedEvent(
                new OrderCreatedEvent(getNextMessageId(), getNextEventId(), customerId, orderId));
    }

    private Order convertCommandIntoOrder(final CreateOrderCommand command) {
        return new Order(command.getCustomerId(), command.getOrderTotal());
    }

    private Order convertDTOIntoOrder(final OrderDTO orderDTO) {
        return new Order(orderDTO.getCustomerId(), orderDTO.getOrderTotal());
    }

    private long saveOrder(final Order order) {
        return outboundPersistencePort.save(order);
    }

    private long getNextMessageId() {
        return RANDOM.nextLong();
    }

    private long getNextEventId() {
        // returned from the saved database event, before sending it (using transactional messaging)
        return RANDOM.nextLong();
    }
}

File: python/src/order_service/services/order_service.py
import logging
from datetime import datetime
from random import randint

import requests
from common.dto.order import OrderDTO
from common.helper.asynch import run_async
from common.message.command.order import (
    ChargeOrderCommand,
    CreateOrderCommand,
    Currency,
)
from common.message.event.order import OrderCreatedEvent
from flask import current_app

from order_service.models import Order
from order_service.repositories.order import OrderRepository


class OrderService:
    def __init__(self):
        self.logger = logging.getLogger(__name__)

    # creating an order received from a REST endpoint (UI, testing app etc)
    def create_order_rest(self, order_dto: OrderDTO):
        order_repository = OrderRepository()

        order = Order(
            customer_id=order_dto.customer_id, total=order_dto.get_order_total(), creation_date=datetime.now()
        )
        order_id = order_repository.save(order)

        self.logger.info(
            f"Creating an order for the customer with the ID {order.customer_id}, for {order.order_items} items..."
        )

        run_async(
            (
                lambda: (
                    self.publish_charge_order(
                        order.customer_id, order_id, order.total
                    ),
                    self.publish_order_created_event(order.customer_id, order_id),
                )
            )
        )

    def publish_charge_order(self, customer_id, order_id, order_total):
        producer = current_app.extensions["producer"]
        producer.publish_charge_order_command(
            ChargeOrderCommand(
                message_id=self.__get_next_message_id(),
                customer_id=customer_id,
                order_id=order_id,
                order_total=order_total,
                currency=Currency.EUR,
            )
        )

    def publish_order_created_event(self, customer_id, order_id):
        producer = current_app.extensions["producer"]
        producer.publish_order_created_event(
            OrderCreatedEvent(
                message_id=self.__get_next_message_id(),
                event_id=self.__get_next_event_id(),
                customer_id=customer_id,
                order_id=order_id,
            )
        )

    # creating an order received from a messaging endpoint (upstream system, 3rd party application etc)
    def create_order_messaging(self, create_order_command: CreateOrderCommand):
        self.logger.info(
            f"Creating an order for the product {create_order_command.product_name}, for the customer with the ID {create_order_command.customer_id}..."
        )

        order = Order(
            customer_id=create_order_command.customer_id,
            total=create_order_command.order_total,
            creation_date=datetime.now()
        )
        order_repository = OrderRepository()

        order_id = order_repository.save(order)

        self.publish_charge_order(order.customer_id, order_id, order.total)
        self.publish_order_created_event(order.customer_id, order_id)


    def charge_order(self, customer_id, order_id):
        # Constructing the URL
        url = f"{current_app.config["BILLING_SERVICE_ENDPOINT"]}/{customer_id}/{order_id}"

        try:
            response = requests.post(url)
            self.logger.info("Got the response: %s", response.text)
            return response

        except requests.RequestException as e:
            self.logger.error("Error occurred while making the request: %s", e)
            return None

    def __get_next_message_id(self):
        # Generate a random message ID (for demonstration purposes)
        return randint(0, 2**64 - 1)

    def __get_next_event_id(self):
        # Generate a random event ID (for demonstration purposes)
        return randint(0, 2**64 - 1)
